### PR TITLE
Kubelet: implement support for podAndContainerStatsFromCRI

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
This PR uses the CRI changes in https://github.com/kubernetes/kubernetes/pull/102789 to implement the feature gate for kubernetes/enhancements#2364  
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Add support for PodAndContainerStatsFromCRI featuregate, which allows a user to specify their pod stats must also come from the CRI, not cAdvisor.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs


- [KEP]: https://github.com/kubernetes/enhancements/issues/2371
```

### CODE REVIEW

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 1]
```diff
@@ -1272,6 +1385,15 @@ message FilesystemUsage {
     // This may not equal InodesCapacity - InodesAvailable because the underlying
     // filesystem may also be used for purposes other than storing images.
     UInt64Value inodes_used = 4;
+    // TODO: Unclear how the remaining fields relate to container stats. Is it filled in cAdvisor?
```
Comment (haircommander) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1388]: also TODO: I have yet to actually figure out how to wire these up yet. it's not clear this should be the responsibility of CRI impl or cAdvisor

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 2]
```diff
@@ -545,6 +551,114 @@ message ListPodSandboxResponse {
...
+    // Filter for the list request.
+    PodSandboxStatsFilter filter = 1;
+}
+
+message ListPodSandboxStatsResponse {
+    // Stats of the pod.
+    repeated PodSandboxStats stats = 1;
+}
+
+// PodSandboxAttributes provides basic information of the container.
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:584]: nit: `of the pod`

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 3]
```diff
@@ -545,6 +551,114 @@ message ListPodSandboxResponse {
...
+    WindowsPodSandboxStats windows = 3;
+}
+
+// LinuxPodSandboxStats provides the resource usage statistics for a pod on linux
+message LinuxPodSandboxStats {
+    // CPU usage gathered from the pod.
+    CpuUsage cpu = 1;
+    // Memory usage gathered from the pod.
+    MemoryUsage memory = 2;
+    // Stats pertaining to CPU resources consumed by pod cgroup (which includes all containers' resource usage and pod overhead).
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:616]: This comment is for `MemoryUsage`. Can you move it up and add right one for network stats?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 4]
```diff
@@ -545,6 +551,114 @@ message ListPodSandboxResponse {
...
+message WindowsPodSandboxStats {
+    // TODO: Add stats relevant to windows
+}
+
+// NetworkStats contains data about network resources.
+message NetworkStats {
+    // The time at which these stats were updated.
+    int64 timestamp = 1;
+
+    // Stats for the default interface, if found
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:633]: Is `if found` needed? (There will always be a default interface, right?)

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 1 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto]: Does this make more sense as a cumulative stat rather than associating with an ImageSpec for a container?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 5]
```diff
@@ -549,6 +555,114 @@ message ListPodSandboxResponse {
...
+    // The time at which these stats were updated.
+    int64 timestamp = 1;
+
+    // Stats for the default interface, if found
+    InterfaceStats default_interface = 2;
+
+    repeated InterfaceStats interfaces = 3;
+}
+
+// InterfaceStats contains resource value data about interface.
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: nit: `about a network interface`

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 6]
```diff
@@ -1392,15 +1392,6 @@ message FilesystemUsage {
     // This may not equal InodesCapacity - InodesAvailable because the underlying
     // filesystem may also be used for purposes other than storing images.
     UInt64Value inodes_used = 4;
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto]: Can we fold this commit into the first one?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 7]
```diff
@@ -614,9 +614,9 @@ message LinuxPodSandboxStats {
     // Memory usage gathered from the pod.
     MemoryUsage memory = 2;
     // Stats pertaining to CPU resources consumed by pod cgroup (which includes all containers' resource usage and pod overhead).
-    NetworkStats network = 3;
+    NetworkUsage network = 3;
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:617]: Can we merge this commit into the first commit?

File: pkg/kubelet/cri/remote/remote_runtime.go [hunk 8]
```diff
@@ -569,6 +569,46 @@ func (r *remoteRuntimeService) ListContainerStats(filter *runtimeapi.ContainerSt
...
+		}
+		return nil, err
+	}
+	r.logReduction.ClearID(podSandboxID)
+	klog.V(10).InfoS("[RemoteRuntimeService] PodSandbox Response", "podSandboxID", podSandboxID, "stats", resp.GetStats())
+
+	return resp.GetStats(), nil
+}
+
+// ListContainerStats returns the list of ContainerStats given the filter.
```
Comment (bobbypage) [inline pkg/kubelet/cri/remote/remote_runtime.go:593]: godoc comment is out of date

```
ListPodSandboxStats returns the list of pod sandbox stats given the filter
```

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 9]
```diff
@@ -830,3 +889,71 @@ func getCRICadvisorStats(infos map[string]cadvisorapiv2.ContainerInfo) map[strin
...
+		return
+	}
+	criCPU := criPodStat.Linux.Cpu
+	ps.CPU = &statsapi.CPUStats{
+		Time:                 metav1.NewTime(time.Unix(0, criCPU.Timestamp)),
+		UsageNanoCores:       valueOfUInt64Value(criCPU.UsageNanoCores),
+		UsageCoreNanoSeconds: valueOfUInt64Value(criCPU.UsageCoreNanoSeconds),
+	}
+
+	if criPodStat == nil || criPodStat.Linux == nil || criPodStat.Linux.Memory == nil {
```
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:930]: we already have the `criPodStat == nil || criPodStat.Linux == nil` checks above on line 920, i think this is redundant

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 10]
```diff
@@ -830,3 +889,71 @@ func getCRICadvisorStats(infos map[string]cadvisorapiv2.ContainerInfo) map[strin
...
+	return statsapi.InterfaceStats{
+		Name:     criIface.Name,
+		RxBytes:  valueOfUInt64Value(criIface.RxBytes),
+		RxErrors: valueOfUInt64Value(criIface.RxErrors),
+		TxBytes:  valueOfUInt64Value(criIface.TxBytes),
+		TxErrors: valueOfUInt64Value(criIface.TxErrors),
+	}
+}
+
+func addCRIPodCPUAndMemoryStats(ps *statsapi.PodStats, criPodStat *runtimeapi.PodSandboxStats) {
```
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:919]: nit: i would separate this two functions, one for cpu stats, one for memory stats

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 11]
```diff
@@ -634,6 +686,10 @@ func (p *criStatsProvider) getContainerUsageNanoCores(stats *runtimeapi.Containe
 		return nil
 	}
 
+	if stats.Cpu != nil && stats.Cpu.UsageNanoCores != nil {
```
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:689]: just to understand, reason we don't need to acquire mutex etc, is because we are not using the "cached" `usageNanoCores` and instead using one directly supplied from the stats message?

Comment (bobbypage) [reply pkg/kubelet/stats/cri_stats_provider.go:689]: is there any reason for us to cache this `UsageNanoCores` in CRI stats world or is it unnecessary now?

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 12]
```diff
@@ -212,45 +204,80 @@ func (p *criStatsProvider) listPodStats(updateCPUNanoCoreUsage bool) ([]statsapi
...
+		summarySandboxStats = append(summarySandboxStats, *ps)
 	}
+	return summarySandboxStats, nil
+}
 
+// ListPodCPUAndMemoryStats returns the CPU and Memory stats of all the pod-managed containers.
+func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, error) {
 	// sandboxIDToPodStats is a temporary map from sandbox ID to its pod stats.
 	sandboxIDToPodStats := make(map[string]*statsapi.PodStats)
+	containerMap, podSandboxMap, err := p.getPodAndContainerMaps()
```
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:253]: just to confirm the reason that the is removed from the diff is because this logic is now handled by `getPodAndContainerMaps`, correct?

```
containers = removeTerminatedContainers(containers)
	// Creates container map.
	containerMap := make(map[string]*runtimeapi.Container)
	for _, c := range containers {
		containerMap[c.Id] = c
	}
```

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 13]
```diff
@@ -295,13 +322,38 @@ func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, erro
...
 	return result, nil
 }
 
+func (p *criStatsProvider) getPodAndContainerMaps() (map[string]*runtimeapi.Container, map[string]*runtimeapi.PodSandbox, error) {
+	containers, err := p.runtimeService.ListContainers(&runtimeapi.ContainerFilter{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to list all containers: %v", err)
+	}
+
+	// Creates pod sandbox map.
```
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:337]: can we note what is key -> value here?

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 14]
```diff
@@ -295,13 +322,38 @@ func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, erro
...
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to list all pod sandboxes: %v", err)
+	}
+	podSandboxes = removeTerminatedPods(podSandboxes)
+	for _, s := range podSandboxes {
+		podSandboxMap[s.Id] = s
+	}
+
+	containers = removeTerminatedContainers(containers)
+	// Creates container map.
```
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:349]: ditto here, note what is what is key -> value mapping

File: pkg/features/kube_features.go [hunk 15]
```diff
@@ -761,6 +761,13 @@ const (
 	//
 	// Allows clients to request a duration for certificates issued via the Kubernetes CSR API.
 	CSRDuration featuregate.Feature = "CSRDuration"
+
+	// owner: @haircommander
+	// kep: http://kep.k8s.io/2364
+	// alpha: v1.22
+	//
+	// Configures the Kubelet to use the CRI to populate pod and container stats, instead of supplimenting with stats from cAdvisor.
```
Comment (bobbypage) [inline pkg/features/kube_features.go:769]: Maybe note that this requires CRI runtime to support pod and container level stats.

File: pkg/kubelet/dockershim/docker_stats.go [hunk 16]
```diff
@@ -62,3 +65,15 @@ func (ds *dockerService) ListContainerStats(ctx context.Context, r *runtimeapi.L
...
 	return &runtimeapi.ListContainerStatsResponse{Stats: stats}, nil
 }
+
+// PodSandboxStats returns stats for a pod stats request based on pod id.
+// This function is not implemented for the dockershim.
+func (ds *dockerService) PodSandboxStats(_ context.Context, r *runtimeapi.PodSandboxStatsRequest) (*runtimeapi.PodSandboxStatsResponse, error) {
+	return nil, ErrNotImplemented
+}
+
+// ListPodSandboxStats returns stats for a list container stats request based on a filter.
```
Comment (bobbypage) [inline pkg/kubelet/dockershim/docker_stats.go:75]: ListPodSandboxStats returns stats for a list pod sandbox stats request...

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 13 above)
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:337]: ditto, add doco on key -> value mapping of this map and below one

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 11 above)
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:689]: add comment to explain why there's two code paths here and why in this case we avoid using cached `UsageNanoCores`

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 17]
```diff
@@ -653,6 +709,9 @@ func (p *criStatsProvider) getAndUpdateContainerUsageNanoCores(stats *runtimeapi
 	if stats == nil || stats.Attributes == nil || stats.Cpu == nil || stats.Cpu.UsageCoreNanoSeconds == nil {
 		return nil
 	}
+	if stats.Cpu.UsageNanoCores != nil {
```
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:712]: ditto here, worth to explain that cache is unused

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 18]
```diff
@@ -545,6 +551,113 @@ message ListPodSandboxResponse {
     repeated PodSandbox items = 1;
 }
 
+message PodSandboxStatsRequest {
```
Comment (bobbypage) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:554]: I assume proto changes will not be included in this PR but rather in #102789 instead correct?

Comment (haircommander) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:554]: yeah this is based on that

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 10 above)
Comment (haircommander) [reply pkg/kubelet/stats/cri_stats_provider.go:919]: yeah i find it weird the cadvisor variants of this don't do that.

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 12 above)
Comment (haircommander) [reply pkg/kubelet/stats/cri_stats_provider.go:253]: yup!

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 13 above)
Comment (haircommander) [reply pkg/kubelet/stats/cri_stats_provider.go:337]: added

Comment (haircommander) [reply pkg/kubelet/stats/cri_stats_provider.go:337]: added

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 17 above)
Comment (haircommander) [reply pkg/kubelet/stats/cri_stats_provider.go:712]: done!

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 11 above)
Comment (haircommander) [reply pkg/kubelet/stats/cri_stats_provider.go:689]: I don't see much need to deal with locking if we're going to bypass the cache now

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 9 above)
Comment (haircommander) [reply pkg/kubelet/stats/cri_stats_provider.go:930]: however, now that they're separated, they are both now needed

**Review by mikebrow — APPROVED** [review]
LGTM
Noting: Per the KEP.. this alpha stage has the feature gate (PodAndContainerStatsFromCRI) off by default so we can discover/work on performance issues and expand testing for when the feature gate gets enabled by default in a later release.

**Comment by haircommander** [comment]
implements many of the alpha pieces of  https://github.com/kubernetes/enhancements/issues/2371

**Comment by rphillips** [comment]
This looks ready.

/lgtm

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 19]
```diff
@@ -122,21 +127,29 @@ func (p *criStatsProvider) listPodStats(updateCPUNanoCoreUsage bool) ([]statsapi
...
+		_, err := p.listPodStatsStrictlyFromCRI(updateCPUNanoCoreUsage, containerMap, podSandboxMap, &rootFsInfo)
+		if err != nil {
+			s, ok := status.FromError(err)
+			// Legitimate failure, rather than the CRI implementation does not support ListPodSandboxStats.
+			if !ok || s.Code() != codes.Unimplemented {
+				return nil, err
+			}
+			// CRI implementation doesn't support ListPodSandboxStats, warn and fallback.
+			klog.V(5).ErrorS(err,
+				"CRI implementation must be updated to support ListPodSandboxStats if PodAndContainerStatsFromCRI feature gate is enabled. Falling back to populating with cAdvisor; this call will fail in the future.",
```
Comment (bobbypage) [inline pkg/kubelet/stats/cri_stats_provider.go:145]: +1 on adding this as a fallback

**Review by bobbypage — APPROVED** [review]

**Comment by bobbypage** [comment]
LGTM as well, thanks @haircommander for your work on this. 

We plan to add new E2E tests that will validate CRI stats using running against updated container runtime implementations that support the new endpoints.

/lgtm

**Comment by derekwaynecarr** [comment]
thanks @haircommander and @bobbypage for the mutual work on this.
/approve
/lgtm
